packages/ui/src/components/branch-selector/branch-selector-dropdown.tsx

```diff
@@ -0,0 +1,189 @@
+import { useState } from 'react'
+import { Link } from 'react-router-dom'
+
+import {
+  Badge,
+  DropdownMenuContent,
+  DropdownMenuItem,
+  Icon,
+  SearchBox,
+  Tabs,
+  TabsList,
+  TabsTrigger,
+  Text
+} from '@/components'
+import { cn } from '@utils/cn'
+
+import { BranchSelectorBranchProps, BranchSelectorItem, BranchSelectorTab } from './types'
+
+const BRANCH_SELECTOR_LABELS = {
+  [BranchSelectorTab.BRANCHES]: {
+    label: 'Branches',
+    searchPlaceholder: 'Find a branch'
+  },
+  [BranchSelectorTab.TAGS]: {
+    label: 'Tags',
+    searchPlaceholder: 'Find a tag'
+  }
+} as const
+
+export interface BranchSelectorDropdownProps {
+  name: string
+  branchList: BranchSelectorItem[]
+  tagList?: BranchSelectorItem[]
+  selectBranch: (branch: string) => void
+}
+
+const filterItems = (items: BranchSelectorBranchProps[] | BranchSelectorItem[], query: string) => {
+  if (!query.trim()) return items
+
+  return items.filter(item => item.name.toLowerCase().includes(query.toLowerCase().trim()))
+}
+
+export const BranchSelectorDropdown = ({ name, branchList, tagList, selectBranch }: BranchSelectorDropdownProps) => {
+  const [activeTab, setActiveTab] = useState<BranchSelectorTab>(BranchSelectorTab.BRANCHES)
+  const [searchQuery, setSearchQuery] = useState('')
+
+  const sourceItems = activeTab === BranchSelectorTab.BRANCHES ? branchList : (tagList ?? [])
```

Comment:
If the active tab is not branches and tagList doesn't exist, the array will be different on every render, meaning it's not dependency-array-safe. Could we change tagList to be required? Ideally we want an empty array of tags rather than no list at all as it simplifies all the logic.